lib/block-supports/duotone.php

```diff
@@ -0,0 +1,138 @@
+<?php
+/**
+ * Duotone block support flag.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Registers the style and colors block attributes for block types that support it.
+ *
+ * @param WP_Block_Type $block_type Block Type.
+ */
+function gutenberg_register_duotone_support( $block_type ) {
+	$has_duotone_support = false;
+	if ( property_exists( $block_type, 'supports' ) ) {
+		$has_duotone_support = _wp_array_get( $block_type->supports, array( 'color', 'duotone' ), false );
+	}
+
+	if ( $has_duotone_support ) {
+		if ( ! $block_type->attributes ) {
+			$block_type->attributes = array();
+		}
+
+		if ( ! array_key_exists( 'style', $block_type->attributes ) ) {
+			$block_type->attributes['style'] = array(
+				'type' => 'object',
+			);
+		}
+	}
+}
+
+/**
+ * Render out the duotone stylesheet and SVG.
+ *
+ * @param  WP_Block_Type $block_type       Block type.
+ * @param  array         $block_attributes Block attributes.
+ * @param  string        $block_content    Rendered block content.
+ *
+ * @return string filtered block content.
+ */
+function gutenberg_render_duotone_support( $block_content, $block ) {
+	$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+
+	$duotone_support = false;
+	if ( $block_type && property_exists( $block_type, 'supports' ) ) {
+		$duotone_support = _wp_array_get( $block_type->supports, array( 'color', 'duotone' ), false );
+	}
+
+	$has_duotone_attribute = isset( $block['attrs']['style']['color']['duotone'] );
+
+	if (
+		! $duotone_support ||
+		! $has_duotone_attribute
+	) {
+		return $block_content;
+	}
+
+	$duotone_attribute = $block['attrs']['style']['color']['duotone'];
+
+	$slug_or_id = isset( $duotone_attribute['slug'] ) ? $duotone_attribute['slug'] : uniqid();
```

Comment:
I took a quick look at the PR and this is what I understand.  

- in the server, we embed a `style` tag and a `svg` tag (caveat: blocks may have to apply this filter to more than one element such as cover that needs two while image needs one):

```html
<style>
.wp-block-class.wp-duotone-filter-$slug_or_id <block_supports_selector_one>,
.wp-block-class.wp-duotone-filter-$slug_or_id <block_supports_selector_two> {
  filter: url("#wp-duotone-filter-$slug_or_id");
}
</style>

<svg id="wp-duotone-filter-$slug_or_id"> filter goes here </svg>
```

- in the client, we attach some duotone data to the style block attribute:

```json
style: {
  duotone: {
    slug: 'slug',
    values: {
        r: 'r',
        g: 'g',
        b: 'b'
    }
  }
}
```

- to make the client work, we also attach a `<DuotoneFilter >` to the block's markup but this isn't saved in the post_content. Everything depends on the server generating this for blocks.

---

If the above is correct, I wonder: can we remove the concept of slug from the data we store in the style attribute of the block and always generate the ID in the server? It seems like the only info we need is the values selected by the user in the client for this to work.

Reply:
I guess the logic for this was that there may be "preset" duotones and "custom" duotones selected by the user. However, it looks like we attach the values in both cases as per https://github.com/WordPress/gutenberg/pull/26752/files?file-filters%5B%5D=.js&file-filters%5B%5D=.md&file-filters%5B%5D=.php&file-filters%5B%5D=.scss#diff-37be335b0d340015a545c960c96de07c7e0ea34ce12b3dd8a1735930a526525aR57

Reply:
Another thought that may be in favor of just storing the values: how a theme can apply a duotone to all blocks of a type via theme.json?

I imagine we want this to work like this (extract of a theme.json a theme may use):

```json
"styles": {
  "core/image": {
    "color": {
      "duotone": {
        "r": "...",
        "g": "...",
        "b": "..."
      }
    }
  }
}
```

Reply:
> I wonder: can we remove the concept of slug from the data we store in the style attribute of the block and always generate the ID in the server?

I've already done that on the parent branch (#26361), but haven't updated this branch yet. Previously I was computing the duotone object on select and using the slug to show which was selected. I now pre-compute all the preset values and does a deep comparison of each of them to show which is selected. It's less computationally efficient this way, but it matches developers' expectations here better.

I'd like to eventually have a better system for preset duotone values, but it didn't seem like a requirement to ship this now—we can do that later.

> I guess the logic for this was that there may be "preset" duotones and "custom" duotones selected by the user. However, it looks like we attach the values in both cases.

Yeah, I wanted to avoid writing the same code in PHP that I already have in JS for converting from the color strings, especially since large parts are relying on `tinycolor` in JS.

> Another thought that may be in favor of just storing the values: how a theme can apply a duotone to all blocks of a type via theme.json?

Is that something we get "for free" by changing it? Wouldn't people just be able to provide the same thing under the `values` key?

Reply:
This also relates to https://github.com/WordPress/gutenberg/pull/26752#discussion_r605540834 it seems  we should avoid storing these precomputed values and only store the raw colors directly, that way the  API of theme.json is simpler and matches the attribute value. I'm fine with duplicated logic to "render" the styles in JS and PHP personally.

Reply:
> Is that something we get "for free" by changing it? Wouldn't people just be able to provide the same thing under the values key?

Duotone support in global styles is going to need custom code: it doesn't map a CSS property like the others (fontSize, etc), so we need to implement that logic as well. The server code needs to be reused both for attaching user-defined duotone filters (this PR) & theme-defined duotone filters (via theme.json).